<?xml version="1.0" encoding="iso-8859-1"?>
<!DOCTYPE HTML>
<html>
<head>
 <script type="text/javascript" src="/site_media/jquery/js/jquery-1.6.1.min.js"></script>

 <style type="text/css">
  #formwrap {
   line-height: 2em;
   background: #eef;
   margin: 10px;
   padding: 10px;
   height: 500px;
  }
  body {
   font-family: "Lucida Grande", "Lucida Sans Unicode", Verdana, Arial, Helvetica, sans-serif;
   font-size: 14px;
  }
  .center { margin-left:auto; margin-right:auto; }
  .help {cursor:help; border-bottom: 1px dotted #A9A9A9}
 </style>

 <script>
  $(function()
  {
    // uncheck barcoded targets & sampleID options (allows values to be set consisently for checked state)
    $('#sampleid').attr('checked',false);
    $('#barcodebeds').attr('checked',false);
    $('.bcbed-shy').hide();
    // turn off async JavaScript so we can easily get/set var names
    $.ajaxSetup({async: false});
    // get the reference for the current report - TS_result is defined in run verison of this script
    var reportUrl = '/rundb/api/v1/results/' + TB_result  + '/?format=json&noplugin=True';
    var referenceID;
    var experimentUrl;
    var analysisUrl;
    $.get( reportUrl, function(data)
    {
     analysisUrl = data.reportLink;
     experimentUrl = data.experiment;
     referenceID = data.reference;
    });
    // attempt to get Plan details for autosetting of GUIs
    var planLibType = "";
    var planBedFile = "";
    if( experimentUrl != undefined && experimentUrl != null && experimentUrl != "" ) {
      var planUrl;
      $.get( experimentUrl+"?format=json", function(data) { planUrl = data.plan; } );
      if( planUrl != undefined && planUrl != null && planUrl != "" ) {
        $.get( planUrl+"?format=json", function(data) {
          planLibType = data.runType;
          planBedFile = data.bedfile;
        });
      }
    }
    document.getElementById('referenceid').innerHTML = referenceID;
    var populateBarcodeTargets = 0;
    var hideMappedBedOption = false; // (referenceID != "hg19");
    if( hideMappedBedOption ) {
      $('#multibed').hide();
    } else {
      // check/get barcodes
      var barcodeUrl = analysisUrl + 'barcodeList.txt';
      $.get( barcodeUrl, function(mem) {
        var lines = mem.split("\n");
        $.each(lines, function(n,row) {
          var fields = $.trim(row).split(',');
          if( fields[0].substr(0,7) == "barcode" ) {
            $("#barcodeids").append("<option value='" + fields[1] + "'>" +  fields[1] + "</option>");
            ++populateBarcodeTargets;
          }
        });
        if( populateBarcodeTargets < 2 ) {
          $('#multibed').hide();
        } else {
          // sort the list (for old reports)
          var lst = $('#barcodeids option');
          lst.sort(function(a,b) {
            if( a.text > b.text ) return 1;
            if( a.text < b.text ) return -1;
            return 0
          });
          $("#barcodeids").empty().append(lst);
          $('#barcodeids option')[0].selected = true;
        }
      }).error(function() {
        $('#multibed').hide();
      });
    }

    var refBedType = '/unmerged/detail/';
    var refBedURL = '/rundb/api/v1/content/?format=json&limit=0&publisher__name=BED&path__startswith=/'+referenceID+refBedType;
    $.get( refBedURL, function(data)
    {
      $.each( data.objects, function(intIndex,result)
      {
       var targfile = result.file;
       var i = targfile.lastIndexOf(refBedType);
       if( i < 0 ) return true;
       var selName = targfile.substr(i+refBedType.length)
       if( selName.substr(0,7) == 'example' ) return true;
       var j = selName.lastIndexOf('.bed');
       if( j > 0 && !result.meta.hotspot )
       {
        selName = selName.substr(0,j);
        $("#targetregions").append("<option value='" + targfile + "'>" +  selName + "</option>");
        if( populateBarcodeTargets > 0 ) {
          $("#barcodetargets").append("<option value='" + targfile + "'>" +  selName + "</option>");
        }
       }
      });
       var lst = $('#targetregions option');
       lst.sort(function(a,b) {
         if( a.text > b.text ) return 1;
         if( a.text < b.text ) return -1;
         return 0
       });
       $("#targetregions").empty().append("<option value=''>None</option>");
       $("#targetregions").append(lst);
       $('#targetregions option')[0].selected = true;
       if( populateBarcodeTargets > 0 ) {
         // have to repeat operation with new list to prevent issues with sharing objects(?)
         lst = $('#barcodetargets option');
         lst.sort(function(a,b) {
           if( a.text > b.text ) return 1;
           if( a.text < b.text ) return -1;
           return 0
         });
         $("#barcodetargets").empty().append(lst);
         $('#barcodetargets option')[0].selected = true;
       }
    });
    // make the Plan auto-selections
    var pluginLibType = "ampliseq";
    if( planLibType != "" ) {
      //pluginLibType = planBedFile == "" ? "wholegenome" : "ampliseq";
      if( planLibType.substr(0,4) != "AMPS" ) {
        alert("Run type not set as AmpliSeq - plugin will run as if it were.");
        //$("#json_result").hide();
        //return;
        planLibType = "AMPS";
      }
      if( planLibType == "WGNM" ) {
        pluginLibType = "wholegenome";
      } else if( planLibType == "AMPS" ) {
        pluginLibType = "ampliseq";
      } else if( planLibType == "TARS" ) {
        pluginLibType = "targetseq";
      } else if( planLibType == "AMPS_RNA" ) {
        pluginLibType = "ampliseq-rna";
      } else if( planLibType == "AMPS_EXOME" ) {
        pluginLibType = "ampliseq-exome";
      }
    }
    if( planBedFile != "" ) {
      $("#targetregions").val(planBedFile);
      if( populateBarcodeTargets > 0 ) {
        $("#barcodetargets").val(planBedFile);
      }
    }

    function setOptionDefaults(libType,refID) {
      $("#librarytype").val(libType);
      if( libType == "ampliseq" ) {
        // Note: this is a captured variable
        if( referenceID == "hg19" ) {
          //$("#sampid").show();
        } else {
          $("#sampid").hide();
        }
        //$("#ptrim").show();
        //$("#tpadding").hide();
        //$("#uniquereads").show();
        //$("#nondups").hide();
      } else if( libType == "targetseq" ) {
        $("#sampid").hide();
        //$("#ptrim").hide();
        //$("#tpadding").show();
        //$("#uniquereads").show();
        //$("#nondups").show();
      } else if( libType == "ampliseq-rna" ) {
        $("#sampid").show();
        //$("#ptrim").hide();
        //$("#tpadding").hide();
        //$("#uniquereads").hide();
        //$("#nondups").hide();
      } else {
        $("#sampid").hide();
        //$("#ptrim").hide();
        //$("#tpadding").show();
        //$("#uniquereads").show();
        //$("#nondups").show();
      }
    }

    // Set default options according to Library type selected (and reference for scope limiting)
    setOptionDefaults(pluginLibType,referenceID);

    // Primer Trimming option disabled until future resolution of undesired side effects
    $("#trimreads").attr("checked",false);

    $("#librarytype").change(function() {
      var libType = $("#librarytype option:selected").val();
      setOptionDefaults(libType);
    });

    $('#barcodebeds').change(function() {
      if( $(this).attr("checked") ) {
        $('.bcbed-shy').show();
      } else {
        $('.bcbed-shy').hide();
      }
    });

    $('#addbcbutton').click(function() {
      var bcID = $('#barcodeids option:selected').text();
      var trgID = $('#barcodetargets option:selected').text();
      var box = $("#bctextarea");
      // clean current input by removing duplicates and blank lines
      var lines = box.val().split("\n");
      var text = "";
      $.each(lines, function(n,row) {
        var fields = $.trim(row).split(/\s+/);
        if( fields.length >= 2 && fields[0] != "" && fields[0] != bcID ) {
          text += fields[0]+'\t'+fields[1]+'\n';
        }
      });
      box.val(text+bcID+'\t'+trgID+'\n');
      var bcIndex = document.AmpliCat.barcodeids.selectedIndex + 1;
      var numOptions = $('#barcodeids option').size();
      if( bcIndex >= numOptions ) bcIndex = 0;
      $('#barcodeids option')[bcIndex].selected = true;
    });

    // Check (user) NPVs are correctly formated and available
    // Return a compressed list of mappings or error message "Error at line ..."
    function checkBarcodeTargets(txtList) {
      var lines = txtList.split("\n");
      var text = "";
      $.each(lines, function(n,row) {
        var fields = $.trim(row).split(/\s+/);
        if( fields.length == 1 && fields[0] == "" ) {
          return true;  // blank lines ok
        } else if( fields.length != 2 || fields[1] == "" ) {
          text = "Error at line "+n+": Invalid number of fields. Must be '<barcode> <targets>'.";
          return false;
        } else if( $('#barcodeids option[value='+fields[0]+']').length == 0 ) {
          text = "Error at line "+n+": Barcode ID "+fields[0]+" is unavailable to this run analysis.";
          return false;
        }
        var trgIndex = 0;
        var inList = false;
        $('#barcodetargets option').each(function() {
          if( this.text == fields[1] ) {
            inList = true;
            return false;
          }
          ++trgIndex;
        });
        if( !inList ) {
          text = "Error at line "+n+": Target Regions "+fields[1]+" is unavailable to this reference.";
          return false;
        }
        // collect the actual file path to the bed file
        text += fields[0]+'='+document.getElementById('barcodetargets')[trgIndex].value+';';
      });
      return text;
    }

    $.fn.serializeObject = function()
    {
     var o = {};
     var a = this.serializeArray();
     $.each(a,
      function() {
       if (o[this.name] != null) {
        if (!o[this.name].push) {
         o[this.name] = [o[this.name]];
        }
        o[this.name].push(this.value || '');
       } else {
        o[this.name] = this.value || '';
       }
      });
     return o;
    };

    $(function()
    {
     $('#postbutton').click(function()
     {
      var libn = document.AmpliCat.librarytype.selectedIndex;
      var libType = document.getElementById('librarytype')[libn].value;
      var libName = document.getElementById('librarytype')[libn].text;
      var tarn = document.AmpliCat.targetregions.selectedIndex;
      var useBcTargets = $('#barcodebeds').attr("checked");
      var useSampleIDs = $('#sampleid').attr("checked");
      var bcTargetList = "";
      if( useBcTargets ) {
        bcTargetList = checkBarcodeTargets( $("#bctextarea").val() );
        if( bcTargetList == "" ) {
          var msg = "Barcode-specific Targets was specified but no mappings were defined."
          msg += "\n\nDo you wish to continue?";
          if( !confirm(msg) ) return false;
        } else if( bcTargetList.substr(0,5) == "Error" ) {
          alert("Barcode target mapping is invalid.\n"+bcTargetList);
          return false;
        }
      }
      if( libType == "wholegenome" ) {
       if( tarn > 0 || bcTargetList != "" ) {
        var msg = "Selecting targeted regions is typically not appropriate for Library Type: " + libName;
        msg += ". (Coverage will be considered in these just regions.)";
        msg += "\n\nDo you wish to continue?";
        if( !confirm(msg) ) return false;
       }
      } else {
       if( tarn == 0 && bcTargetList == "" ) {
        alert("Targeted Regions selection is required for Library Type: " + libName);
        return false;
       }
      }
      // check for appropriate padding values
      if( libType == "ampliseq" ) {
        document.AmpliCat.padtargets.value = 0;
      }
      var regex = /^[0-9]+$/;
      var padval = document.AmpliCat.padtargets.value;
      if( !regex.test(padval) ) {
       alert("Target padding value must be a non-negative integer (>=0).");
       return false;
      }
      if( tarn == 0 && padval > 0 && bcTargetList == "" ) {
       var msg = "Target padding is ignored if no specific Targeted Regions value is selected.";
       msg += "\n\nDo you wish to continue?";
       if( !confirm(msg) ) return false;
       padval = document.AmpliCat.padtargets.value = 0;
      }
      // safety code: targets are required for AmpliSeq runs
      if( useSampleIDs && tarn == 0 && bcTargetList == "" ) {
       var msg = "SampleID Tracking is typically used to account for additional targets in AmpliSeq runs.";
       msg += "\n(It is expected that Targeted Regions selection is made.)";
       msg += "\nTo look at coverage for just tracking targets it is recommended to use the sampleID plugin.";
       msg += "\n\nDo you wish to continue?";
       if( !confirm(msg) ) return false;
      }
      // force invisible options to be unchecked
      if( libType == "wholegenome" ) {
       //$("#trimreads").attr("checked",false);
       $("#padtargets").attr("value","0");
      } else if( libType == "ampliseq" ) {
       $("#nonduplicates").attr("checked",false);
       $("#padtargets").attr("value","0");
      } else if( libType == "targetseq" ) {
       //$("#trimreads").attr("checked",false);
      } else if( libType == "ampliseq-rna" ) {
       //$("#trimreads").attr("checked",false);
       $("#padtargets").attr("value","0");
       $("#nonduplicates").attr("checked",false);
       $("#uniquereads").attr("checked",false);
      }
      // pass user selection with spaces translated to underscores
      var usrSel = document.getElementById('librarytype')[libn].text;
      document.AmpliCat.librarytype_id.value = usrSel.replace(/\s/g,'_');
      usrSel = document.getElementById('targetregions')[tarn].text;
      document.AmpliCat.targetregions_id.value = usrSel.replace(/\s/g,'_');
      document.AmpliCat.barcodetargetregions.value = bcTargetList;
      obj = $('#AmpliCat').serializeObject();
      pluginAPIJSON = { "plugin" : [TB_plugin.fields.name], "pluginconfig" : obj };
      pluginAPIJSON = JSON.stringify(pluginAPIJSON);
      pluginURL = "/rundb/api/v1/results/" + TB_result + "/plugin/";
      $.ajax({
       type: 'POST',
       url: pluginURL,
       async: false,
       contentType: "application/json; charset=utf-8",
       success: function(data)
       {
        $("#json_result").html('<div style="text-align: center;"><img src="/site_media/jquery/colorbox/images/loading.gif" alt="Running Plugin" style="float:center"></img><p>Running the Plugin... Check AmpliCat.html in the Plugin Summary</p></div>');
        setTimeout("parent.$.fn.colorbox.close()",2000);
       },
       data: pluginAPIJSON,
       dataType: "json"
      });
     });
    });

  });
 </script>
</head>

<body>
 <form id="AmpliCat" name="AmpliCat">
  <div style="text-align:center">
   <h1>AmpliCat Plugin</h1>
   <h3>AmpliSeq On-target Read Catagorization</h3>
  </div>
   <table class="center" cellpadding="5px">
    <tr>
     <td><span class="help" title="The short name of the reference genome (or DNA sequences) that the current report was generated against and to be used for coverage analysis.">Reference Genome:</span></td>
     <td><div id="referenceid"></div></td> </tr>
    <tr style="display:none">
     <td><span class="help" title="Select the library (enrichment) type sequenced.">Library Type:</span></td>
     <td><select id="librarytype" name="librarytype">
      <option value="wholegenome">Whole Genome</option>
      <option value="ampliseq" selected="selected">Ion AmpliSeq</option>
      <option value="targetseq">Ion TargetSeq</option>
      <option value="ampliseq-rna">Ion AmpliSeq RNA</option>
     </select></td> </tr>
    <tr>
     <td><span class="help"
       title="Select the target regions (ROI) matching your reference and enriched fragment library. See Description and Usage Notes below.">
       Targeted Regions:</span></td>
     <td><select id="targetregions" name="targetregions"></select></td> </tr>
    <tr id="multibed">
     <td><span class="help"
       title="Assign target regions based on assigned barcodes.">
       Barcode-specific Targets:</span></td>
     <td><input type="checkbox" id="barcodebeds" name="barcodebeds" checked="checked" value="Yes" style="float:left"/>
       <div class="bcbed-shy" style="margin-left:25px">
         Use the text box below to specify barcode to target regions mappings.<br/>
         Hint: Use the 'Add' button and barcode/target selectors for initial set up.<br/>
         Copy the text generated and then use copy/paste for quick entry next time.<br/></div>
       <div class="bcbed-shy">
         <input id="addbcbutton" type="button" value="Add" style="font-size:13px">
         <select id="barcodeids" name="barcodeids" style="font-size:13px"></select>
         <select id="barcodetargets" name="barcodetargets" style="font-size:13px"></select>
       </div>
       <div class="bcbed-shy">
         <textarea id="bctextarea" rows=17 cols="60" style="font-size:13px;border:5px double #C0C0C0"></textarea>
       </div>
       <div class="bcbed-shy" style="font-style:italic;max-width:540px">
         Un-assigned barcodes will use the Targeted Regions selected when the Submit button is clicked.
         (Set to 'None' to ignore un-assigned barcodes.)</div>
     </td> </tr>
    <tr id="sampid" style="display:none">
     <td><span class="help"
       title="Check if this AmpliSeq library employed sampleID tracking amplicons.">
       SampleID Tracking:</span></td>
     <td><input type="checkbox" id="sampleid" name="sampleid" checked="checked" value="Yes"/></td> </tr>
    <tr id="ptrim" style="display:none">
     <td><span class="help"
       title="Trim reads to amplicon targets. See Description and Usage Notes below.">
       Trim Reads:</span></td>
     <td><input type="checkbox" id="trimreads" name="trimreads" checked="checked" value="Yes"/></td> </tr>
    <tr id="tpadding" style="display:none">
     <td><span class="help"
       title="Specifiy the number of bases for option target padding. See Description and Notes below.">
       Target Padding:</span></td>
     <td><input type="text" size=6 id="padtargets" name="padtargets" value=0></select></td> </tr>
    <tr id="uniquereads" style="display:none">
     <td><span class="help"
       title="Check to analyze only reads mapped to a specific sequence in the genome. See Description and Usage Notes below.">
       Use Only Uniquely Mapped Reads:</span></td>
     <td><input type="checkbox" id="uniquemaps" name="uniquemaps" value="Yes" checked="checked"/></td> </tr>
    <tr id="nondups" style="display:none">
     <td><span class="help"
       title="Check to analyze only reads that are not demarked as duplicate reads. See Description and Usage Notes below.">
       Use Only Non-duplicate Reads:</span></td>
     <td><input type="checkbox" id="nonduplicates" name="nonduplicates" value="Yes" checked="checked"/>
     &nbsp;(Requires alignments with Mark Duplicates applied.)</td> </tr>
   </table>
   <br/>
   <div align="center" id="json_result">
    <input id="postbutton" type="submit" value="Submit">
   </div>
   <input type="hidden" id="librarytype_id" name="librarytype_id"/>
   <input type="hidden" id="targetregions_id" name="targetregions_id"/>
   <input type="hidden" id="barcodetargetregions" name="barcodetargetregions" value=""/>
   <br/>
   <div align="left">
    <h3>Description and Usage Notes</h3>
    <p>
This plugin analyzes on-target AmpliSeq panel reads for certain catagories of reads and insert coverage phenomina.
   </p>
<!--
    <p>
The Library Type should be set to the type of enriched fragment library used for the run (aka 'Run Type')
and the Targeted Regions selected should correspond to the enriched regions of the reference (if any).
Some options are only presented as available for the Library Type selected, and coverage will only be
reported over the specified target regions. Typically a Targeted Regions selection is required
for enriched library types and set to 'None' for Whole Genome runs.
(The default selections will match those defined by the Run Plan for this report.)</p>
-->
    <p>
If the option for Barcode-specific Targets is presented you may check this option to use and further specify Targeted Regions
for analysis of reads associated with specific barcode IDs.</p>
<!--
    <p>
The SampleID Tracking option is recommended if your AmpliSeq targets (amplicons) were 'spiked' with SampleID sample tracking
targets. This will inform the analysis to produce an extra statistic for reads mapped to these targets so that
the true level of 'off-target' reads is accurately represented. Note that trimming and padding options do not apply to sample
tracking targets.</p>
    <p>
The Trim Reads option is recommended for AmpliSeq panels that 'tile' target regions. It specifies to trim all aligned reads
to the overlaps with their assigned amplicon targets to avoid inclusion of (partially digested) primer sequence in
base coverage statistics. This is to better reflect the actual base coverage considered for variant calling. </p>
    <p>
The Target Padding value is only available for the Ion TargetSeq and the Whole Genome Library Type.
This specifies a number of bases to expand each individual target by at both 3' and 5' ends and is 
applied to the selected target regions before coverage analysis is performed. A setting more than 0 is
not recommended for standard coverage analysis but a value of 100 or 150 may be useful for determining
the degree of near-target base reads.</p>
    <p>
The 'Uniquely Mapped Reads' and 'Non-duplicate Reads' options are used to specify coverage analysis on a filtered
subset of reads. Non-uniquely mapped reads are those that could have equally well aligned to multiple locations
in the whole genome and are typically mapped randomly to one of these locations. Hence, such reads are typically
no use for unambiguously locating variants and may result in a lower estimation of coverage for enriched libraries.
For whole genome libraries these reads may still be an accurate estimation of coverage over repeated regions.
Non-duplicate reads are those not marked as duplicates in the aligned reads (BAM) file and therefore the option
to use these is only effective if duplicate reads were so marked during the alignment process. This option is not
available for Ion AmpliSeq since (PCR) duplicate reads are anticipated and a necessary requirement for coverage.</p>
-->
   </div>
 </form>
</body>
</html>

